MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java

```diff
@@ -0,0 +1,164 @@
+     */
+    private void readFromExtractor(ByteBuffer inputBuffer, int bufferIndex) {
+        int sampleSize = lowResExtractor.readSampleData(inputBuffer, 0);
+        if (sampleSize < 0) {
+            lowResDecoder.queueInputBuffer(
+                    bufferIndex,
+                    0,
+                    0,
+                    0,
+                    MediaCodec.BUFFER_FLAG_END_OF_STREAM
+            );
+        } else {
+            lowResDecoder.queueInputBuffer(
+                    bufferIndex,
+                    0,
+                    sampleSize,
+                    lowResExtractor.getSampleTime(),
+                    0
+            );
+            lowResExtractor.advance();
+        }
+    }
+
+    /**
+     * Retrieve the information of the next available output buffer.
+     * @return the bundle containing the information of the next available output buffer, or null if
+     * the poll call results in a timeout.
+     */
+    private Bundle getAvailableOutputBufferData() {
+        Bundle bufferData = null;
+        try {
+            bufferData = availableOutputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Log.e(TAG, "No output buffer available", e);
+        }
+        return bufferData;
+    }
+
+    /**
+     * Handle calls to nextFrame() and seekTo() by the MotionPhotoReader.
+     * @param messageData A Bundle containing relevant fields from a call to nextFrame() or
+     *                    seekTo().
+     *
+     *                    The fields in the message bundle are:
+     *                    - MESSAGE_KEY: an int indicating which method call this message comes from.
+     *                    - TIME_US: a long containing the timestamp to seek to (seekTo() calls only)
+     *                             in microseconds.
+     *                    - MODE: an int containing the seeking mode.
+     *
+     *                    The fields in the output buffer data bundle are:
+     *                    - TIMESTAMP_US: a long containing the timestamp assigned to the output
+     *                                  buffer, in microseconds.
+     *                    - BUFFER_INDEX: an int containing the index of the output buffer.
+     *
+     */
+    @RequiresApi(api = LOLLIPOP)
+    public void process(Bundle messageData) {
+        int key = messageData.getInt("MESSAGE_KEY");
+        Bundle bufferData;
+        int bufferIndex;
+        long timestamp;
+        switch (key) {
+            case MotionPhotoReader.MSG_NEXT_FRAME:
+                // Get the next available input buffer and read frame data
+                bufferIndex = getAvailableInputBufferIndex();
+                ByteBuffer inputBuffer = lowResDecoder.getInputBuffer(bufferIndex);
+                readFromExtractor(inputBuffer, bufferIndex);
+
+                // Get the next available output buffer and release frame data
+                bufferData = getAvailableOutputBufferData();
```

Comment:
what happens if `getAvailableOutputBufferData()` returns `null` (e.g. takes too long to decode or requires subsequent input frames before being able to decode) ? I'm not sure if it's possible for decoders to do block decoding.. e.g., feed in N frames and then get N outputs after N frames fed in.

Reply:
(and we can deal with the block decoding problem in a subsequent iteration)

Reply:
added todo for separate CL.